### PR TITLE
Add support for emitting private swiftinterface

### PR DESCRIPTION
**Pull request by luispadron**
This PR adds a new feature: `swift.emit_private_swiftinterface` which enables emitting `.private.swiftinterface` files for Swift modules. These interface files are required when using `@_spi` and library evolution.

## Example

```
bazel build //examples/apple/private_swiftinterface    
```

```
Target //examples/apple/private_swiftinterface:private_swiftinterface up-to-date:
  bazel-bin/examples/apple/private_swiftinterface/PrivateSwiftInterfaceLib.swiftdoc
  bazel-bin/examples/apple/private_swiftinterface/PrivateSwiftInterfaceLib.swiftinterface
  bazel-bin/examples/apple/private_swiftinterface/PrivateSwiftInterfaceLib.private.swiftinterface
  bazel-bin/examples/apple/private_swiftinterface/PrivateSwiftInterfaceLib.swiftmodule
  bazel-bin/examples/apple/private_swiftinterface/PrivateSwiftInterfaceLib.swiftsourceinfo
  bazel-bin/examples/apple/private_swiftinterface/libprivate_swiftinterface.a
```

### CODE REVIEW

File: swift/internal/compiling.bzl [hunk 1]
```diff
@@ -2964,6 +2983,20 @@ def _declare_compile_outputs(
     else:
         swiftinterface_file = None
 
+    if are_all_features_enabled(
+        feature_configuration = feature_configuration,
+        feature_names = [
+            SWIFT_FEATURE_ENABLE_LIBRARY_EVOLUTION,
```
Comment (brentleyjones) [inline swift/internal/compiling.bzl:2989]: Should be `SWIFT_FEATURE_SUPPORTS_LIBRARY_EVOLUTION`. Though I wonder if we should have `SWIFT_FEATURE_ENABLE_LIBRARY_EVOLUTION` as well (and add it to the non-private one as well)?

Comment (brentleyjones) [reply swift/internal/compiling.bzl:2989]: Actually, I think you can generate the interface file without library evolution enabled? Which is what the original version supported? If so, I would revert the comment change for the base feature.

File: swift/internal/swift_library.bzl [hunk 2]
```diff
@@ -228,6 +233,7 @@ def _swift_library_impl(ctx):
         module_context.clang.precompiled_module,
         module_context.swift.swiftdoc,
         module_context.swift.swiftinterface,
+        module_context.swift.private_swiftinterface,
```
Comment (brentleyjones) [inline swift/internal/swift_library.bzl:236]: I think we need the same thing in `swift_module_alias`.

Do we need to change something for `swift_library_artifact_collector` or `compile_module_interface_action_configs` as well?

File: swift/internal/compiling.bzl (hunk 1 above)
Comment (luispadron) [reply swift/internal/compiling.bzl:2989]: > Should be `SWIFT_FEATURE_SUPPORTS_LIBRARY_EVOLUTION`. Though I wonder if we should have `SWIFT_FEATURE_ENABLE_LIBRARY_EVOLUTION` as well (and add it to the non-private one as well)?

We have both already it seems:

```
# If enabled, enables features that require the library evolution compilation
# mode. If disabled, features that require library evolution mode are noops.
SWIFT_FEATURE_SUPPORTS_LIBRARY_EVOLUTION = "swift.supports_library_evolution"

# If enabled, requests the `-enable-library-evolution` swiftc flag which is
# required for newer features like swiftinterface file generation.
SWIFT_FEATURE_ENABLE_LIBRARY_EVOLUTION = "swift.enable_library_evolution"
```

That makes sense and I see the reason for split. I'll revert the comment change. 


Side note: Does that mean that `--features="swift.emit_swiftinterface` implies `--features="swift.enable_library_evolution"`?

Comment (luispadron) [reply swift/internal/compiling.bzl:2989]: > Does that mean that --features="swift.emit_swiftinterface implies --features="swift.enable_library_evolution"?

Seems like yes

```starlark
    if ctx.attr._config_emit_swiftinterface[BuildSettingInfo].value:
        extra_features.append(SWIFT_FEATURE_ENABLE_LIBRARY_EVOLUTION)
        extra_features.append(SWIFT_FEATURE_EMIT_SWIFTINTERFACE)
```

File: swift/internal/swift_library.bzl (hunk 2 above)
Comment (luispadron) [reply swift/internal/swift_library.bzl:236]: Added it there and in the `swift_proto_library` as well

**Comment by luispadron** [comment]
I'll work on adding a test with `swift_import` that uses the private interface now

File: swift/internal/compiling.bzl (hunk 1 above)
Comment (brentleyjones) [reply swift/internal/compiling.bzl:2989]: Only in that case. Normally it doesn't.

**Comment by luispadron** [comment]
Tests added!

File: test/rules/swift_library_artifact_collector.bzl [hunk 3]
```diff
@@ -97,6 +109,7 @@ swift_library_artifact_collector = rule(
         "static_library": attr.output(mandatory = False),
         "swiftdoc": attr.output(mandatory = False),
         "swiftinterface": attr.output(mandatory = False),
+        "private_swiftinterface": attr.output(mandatory = False),
```
Comment (brentleyjones) [inline test/rules/swift_library_artifact_collector.bzl:112]:
```suggestion
        "private_swiftinterface": attr.output(mandatory = False),
        "static_library": attr.output(mandatory = False),
        "swiftdoc": attr.output(mandatory = False),
        "swiftinterface": attr.output(mandatory = False),
```

File: test/rules/swift_library_artifact_collector.bzl [hunk 4]
```diff
@@ -84,6 +90,12 @@ def _swift_library_artifact_collector_impl(ctx):
             source = swift_info.direct_modules[0].swift.swiftinterface,
             destination = ctx.outputs.swiftinterface,
         )
+    if ctx.outputs.private_swiftinterface:
+        _copy_file(
+            ctx.actions,
+            source = swift_info.direct_modules[0].swift.private_swiftinterface,
+            destination = ctx.outputs.private_swiftinterface,
+        )
```
Comment (brentleyjones) [inline test/rules/swift_library_artifact_collector.bzl:98]: nit: can you move this up to be alphabetical?

File: test/rules/swift_library_artifact_collector.bzl [hunk 5]
```diff
@@ -24,16 +24,22 @@ load(
...
+        "@build_bazel_rules_swift//swift:emit_private_swiftinterface": attr.private_swiftinterface != None,
     }
 
 _swiftinterface_transition = transition(
     implementation = _swiftinterface_transition_impl,
     inputs = [],
-    outputs = ["@build_bazel_rules_swift//swift:emit_swiftinterface"],
+    outputs = [
+        "@build_bazel_rules_swift//swift:emit_swiftinterface",
+        "@build_bazel_rules_swift//swift:emit_private_swiftinterface",
```
Comment (brentleyjones) [inline test/rules/swift_library_artifact_collector.bzl:41]:
```suggestion
        "@build_bazel_rules_swift//swift:emit_private_swiftinterface",
        "@build_bazel_rules_swift//swift:emit_swiftinterface",
```

File: test/rules/swift_library_artifact_collector.bzl [hunk 6]
```diff
@@ -24,16 +24,22 @@ load(
...
 def _swiftinterface_transition_impl(_settings, attr):
-    # If the `.swiftinterface` file is requested, apply the setting that causes
-    # the rule to generate it.
     return {
+        # If the `.swiftinterface` file is requested, apply the setting that causes
+        # the rule to generate it.
         "@build_bazel_rules_swift//swift:emit_swiftinterface": attr.swiftinterface != None,
+        # If the `.private.swiftinterface` file is requested, apply the setting that causes
+        # the rule to generate it.
+        "@build_bazel_rules_swift//swift:emit_private_swiftinterface": attr.private_swiftinterface != None,
```
Comment (brentleyjones) [inline test/rules/swift_library_artifact_collector.bzl:33]: nit: can you put this above the other, to keep the keys alphabetical?

File: swift/internal/derived_files.bzl [hunk 7]
```diff
@@ -350,6 +362,7 @@ derived_files = struct(
     swiftc_derived_output_file_map = _swiftc_derived_output_file_map,
     swiftdoc = _swiftdoc,
     swiftinterface = _swiftinterface,
+    private_swiftinterface = _private_swiftinterface,
```
Comment (brentleyjones) [inline swift/internal/derived_files.bzl:365]: nit: can you move this up to keep it alphabetical?

File: swift/internal/compiling.bzl [hunk 8]
```diff
@@ -3116,6 +3149,7 @@ def _declare_compile_outputs(
         derived_files_output_file_map = derived_files_output_file_map,
         swiftdoc_file = swiftdoc_file,
         swiftinterface_file = swiftinterface_file,
+        private_swiftinterface_file = private_swiftinterface_file,
```
Comment (brentleyjones) [inline swift/internal/compiling.bzl:3152]: nit: can you move this up to keep it alphabetical?

File: swift/internal/compiling.bzl [hunk 9]
```diff
@@ -2630,6 +2648,7 @@ to use swift_common.compile(include_dev_srch_paths = ...) instead.\
             plugins = depset(plugins),
             swiftdoc = compile_outputs.swiftdoc_file,
             swiftinterface = compile_outputs.swiftinterface_file,
+            private_swiftinterface = compile_outputs.private_swiftinterface_file,
```
Comment (brentleyjones) [inline swift/internal/compiling.bzl:2651]: nit: can you move this up to keep it alphabetical?

File: swift/internal/attrs.bzl [hunk 10]
```diff
@@ -189,6 +189,9 @@ def swift_config_attrs():
         "_config_emit_swiftinterface": attr.label(
             default = "@build_bazel_rules_swift//swift:emit_swiftinterface",
         ),
+        "_config_emit_private_swiftinterface": attr.label(
+            default = "@build_bazel_rules_swift//swift:emit_private_swiftinterface",
+        ),
```
Comment (brentleyjones) [inline swift/internal/attrs.bzl:194]: nit: can you move this up to keep it alphabetical?

**Review by brentleyjones — APPROVED** [review]

**Pull request merged**